Adding an interactive login experience

lostintangent opened the pull request:
This PR introduces an interactive login experience (using the `az-login` component I'm working on), that allows devs to deploy/manage their Azure functions, without needing to explicitly create a service principal and then set the four env vars. This results in a simpler, less error-prone getting started experience. Additionally, this PR results in more code being removed, then added, which is always nice :)

With this change, if those env vars are set, then they will continue to be used (although it also supports the same env vars as Terraform as well, for increased interop). However, if they aren't set, then the CLI will initiate an interactive login by launching the user's default browser to the device login page (`https://aka.ms/devicelogin`), and copying the device code to the clipboard. Once login has been successfully completed, it will create a service principal behind the scenes, and persist in to disk, so that subsequent commands will naturally pick it up.

<img width="888" alt="screen shot 2017-04-25 at 5 10 16 pm" src="https://cloud.githubusercontent.com/assets/116461/25412806/60cd5d28-29da-11e7-968e-6e567cf1c2c4.png">

As far as subscription selection goes, if the authenticated account only has a single subscription, it will simply select that one (preventing the user from needing to set a subscription ID env var). If there are multiple subscriptions, and the user is using the Azure CLI 2.0, it will look to see what their default subscription has been set to from there. If that doesn't exist, then it will fall back to prompting the user for the subscription they would like to use.

<img width="548" alt="screen shot 2017-04-25 at 5 33 07 pm" src="https://cloud.githubusercontent.com/assets/116461/25413233/a314f97c-29dd-11e7-919a-b261e667f173.png">

This way, both the authentication and subscription selection process can be entirely guided.

lostintangent commented:
@pragnagopa @fiveisprime I did a bunch more testing, and everything looks good. Let me know what you think!

pragnagopa commented on provider/azureProvider.js:
```diff
@@ -112,16 +92,14 @@ return this.parsedBindings;
-      });
+  Login() {
+    return login({ interactiveLoginHandler: (code) => {
+        this.serverless.cli.log(`Paste this code (copied to your clipboard) into the launched browser, and complete the authentication process: ${code}`);
```
Can you add more explicit logging to indicate that this is going to create a service principal behind the scenes as new Azure customers will not be familiar with SP user.

lostintangent replied:
I agree that new Azure customers wouldn't be familiar with SPs, which is why this PR doesn't currently log out that it's creating one. The auto-generated SP is meant to be an implementation detail that simply provides the ability to login interactively once, and then continue to use the CLI without needing to login again. I'd personally vote not to add any logging here, because it might just complicate the experience.

pragnagopa replied:
Does ServicePrincipal user created here show up in [portal](https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-group-create-service-principal-portal)? If yes, I think we should let users know about creating SP user.

fiveisprime commented:
I can't get this installed. I get an error that the az-login version isn't available.

I was testing via `$ npm i lostintangent/serverless-azure-functions` 

Did you miss publishing `az-login@0.1.7`?

lostintangent commented:
@fiveisprime Apologies! Published. Try again please :)

pragnagopa commented on provider/azureProvider.js:
```diff
@@ -75,21 +65,11 @@ return constants.providerName;
 
     return new BbPromise((resolve, reject) => {
-        subscriptionId = this.getSetting(azureCredentials.azureSubId);
-        servicePrincipalTenantId = this.getSetting(azureCredentials.azureServicePrincipalTenantId);
```
Remove getSetting function as well

lostintangent replied:
Done!

pragnagopa commented on provider/azureProvider.js:
```diff
@@ -102,27 +82,14 @@ return constants.providerName;
-  }
+  Login() {
+    return login({ interactiveLoginHandler: (code) => {
+        this.serverless.cli.log(`Paste this code (copied to your clipboard) into the launched browser, and complete the authentication process: ${code}`);
```
Does SP user created here show up in [portal](https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-group-create-service-principal-portal)? If yes, I think we should log indicating SP user being created.

lostintangent replied:
That implies users are going to go into the portal and explore their `Azure Active Directory` settings (is that really that common for new Azure users?) Can't we just keep the logging simple and expand it based on feedback? If we log super verbose messages, we run the risk of making it overwhelming/overcomplicated for everyone, as opposed to accommodating the few that would care to know a SP is being created.

I actually think the current logging for the `deploy` is already way too verbose as it is.

pragnagopa replied:
As long as it is mentioned somewhere that SP user is being created automatically, I am ok with not logging. Mentioning it in ReadMe is works!

pragnagopa replied:
Logs can updated to 'verbose' so that more logs show up only when looking for verbose logs. This will default logs simpler.

pragnagopa commented:
@lostintangent - Please update readme as well. Remove all the info that about setting up environment variables.

lostintangent commented:
@pragnagopa Just updated the `README`, let me know what you think. It's quite a bit simpler :)

I'll update the Serverless docs after this PR is merged.

pragnagopa commented on README.md:
```diff
@@ -1,129 +1,75 @@
-1. `cd my-app`
-2. `npm install`
+1. Node.js v6.5.0+ *(this is the runtime version supported by Azure Functions)*
+2. Serverless CLI. You can run `npm i -g serverless` if you don't already have it.
```
serverless-azure-functions needs to be installed before running command: serverless create -t azure-nodejs --name <APP_NAME>

lostintangent replied:
This "create" support is part of the core serverless CLI. Also, the "Serverless-Azure-Functions" module is listed as an NPM dependency of the generated app, so running "npm install" will install it.

pragnagopa replied:
I just tried this and it fails with error:  Cannot find module 'serverless-azure-functions'

lostintangent replied:
@pragnagopa What version of the Serverless CLI do you have installed? It requires `1.9.0`+, which is when the template was released. I just tried it again and it worked for me.

pragnagopa commented on README.md:
````diff
@@ -1,129 +1,75 @@
+    ```bash
+    serverless deploy
     ```
 
````
Add command to deploy one function : serverless deploy function -f <functionName>

pragnagopa commented on README.md:
```diff
@@ -1,129 +1,75 @@
-6. Set up environment variables
-
-     You need to set up environment variables for your subscription id, tenant id, service principal name, and password. 
+3. View the output logs of your newly deployed function:
```
Change to : Stream the Function Logs. We have an [issue ](https://github.com/serverless/serverless-azure-functions/issues/8)tracking fetching logs for an invocation.

pragnagopa commented on README.md:
```diff
@@ -1,129 +1,75 @@
 
-We'll set up an Azure Subscription and our service principal. You can learn more in the [credentials doc]( https://www.serverless.com/framework/docs/providers/azure/guide/credentials).
+1. Create a new service using the standard Node.js template: `serverless create -t azure-nodejs --name <APP_NAME>`
+2. CD into the generated app directory: `cd <APP_NAME>`
```
creating a template is not generating App_Name directory. Remove this for now. This should be fixed in core serverless CLI

lostintangent replied:
Actually, I just need to change this sample to use `-p` instead, and it will create the directory as expected. This was a doc bug.

lostintangent commented:
@pragnagopa I addressed all of your feedback, and re-tested the `serverless create -t azure-nodejs` and can't repro the issue you mentioned. Let me know if you're using Serverless CLI 1.9.0+ and seeing this problem, and I'll look into it again.

mamaso commented:
There may be issues with multiple subscriptions:

![image](https://cloud.githubusercontent.com/assets/9870694/25728507/68e340b6-30e4-11e7-9346-24ac39ca0696.png)

pragnagopa commented:
@mamaso - this PR is merged now. Can you open an issue with repro steps to track issue with  multiple subscriptions?

lostintangent commented:
@mamaso Could you check whether you have the `azureSubId` or `ARM_SUBSCRIPTION_ID` env variable set? And if so, could you validate that it is set to an ID of a subscription that is associated with the identity you authenticated with?
